clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslServerAuthenticator.java

```diff
@@ -232,44 +250,54 @@ public void authenticate() throws IOException {
             return;
         }
 
-        if (netInBuffer == null) netInBuffer = new NetworkReceive(MAX_RECEIVE_SIZE, connectionId);
-
-        netInBuffer.readFrom(transportLayer);
-
-        if (netInBuffer.complete()) {
-            netInBuffer.payload().rewind();
-            byte[] clientToken = new byte[netInBuffer.payload().remaining()];
-            netInBuffer.payload().get(clientToken, 0, clientToken.length);
-            netInBuffer = null; // reset the networkReceive as we read all the data.
-            try {
-                switch (saslState) {
-                    case HANDSHAKE_OR_VERSIONS_REQUEST:
-                    case HANDSHAKE_REQUEST:
-                        handleKafkaRequest(clientToken);
-                        break;
-                    case INITIAL_REQUEST:
-                        if (handleKafkaRequest(clientToken))
-                            break;
-                        // For default GSSAPI, fall through to authenticate using the client token as the first GSSAPI packet.
-                        // This is required for interoperability with 0.9.0.x clients which do not send handshake request
-                    case AUTHENTICATE:
-                        handleSaslToken(clientToken);
-                        // When the authentication exchange is complete and no more tokens are expected from the client,
-                        // update SASL state. Current SASL state will be updated when outgoing writes to the client complete.
-                        if (saslServer.isComplete())
-                            setSaslState(SaslState.COMPLETE);
-                        break;
-                    default:
+        boolean saslHandshakeReauthenticationInput = saslHandshakeReauthenticationReceive != null
+                && netInBuffer == saslHandshakeReauthenticationReceive;
```

Comment:
Can't we set buffers to a good state from `reauthenticate()` so that this method doesn't have to do anything special for re-authentication? We have received a handshake request, we just need to continue just as with authentication?

Reply:
👍 Good point.  I split out the `processPayload()` method to address a cyclomatic complexity issue, and I did not see the simplification at that time.  Done.